content/news/2023-11-04-bevy-0.12/index.md

```diff
@@ -0,0 +1,3116 @@
+}
+```
+
+You can now mutably iterate all entities and access arbitrary components within them:
+
+```rust
+for mut entity in world.iter_entities_mut() {
+    let mut transform = entity.get_mut::<Transform>().unwrap();
+    transform.translation.x += 2.0;
+}
+```
+
+This required reducing the access scope of [`EntityMut`] to _only_ the entity it accesses (previously it had escape hatches that allowed direct [`World`] access). Use [`EntityWorldMut`] for an equivalent to the old "global access" approach.
+
+[`EntityMut`]: https://docs.rs/bevy/0.12.0/bevy/ecs/world/struct.EntityMut.html
+[`EntityWorldMut`]: https://docs.rs/bevy/0.12.0/bevy/ecs/world/struct.EntityWorldMut.html
+[`World`]: https://docs.rs/bevy/0.12.0/bevy/ecs/world/struct.World.html
+
+## Unified configure_sets API
+
+<div class="release-feature-authors">authors: @geieredgar</div>
+
+Bevy's [Schedule-First API](/news/bevy-0-11/#schedule-first-ecs-apis) introduced in **Bevy 0.11** unified most of the ECS scheduler API surface under a single `add_systems` API. However, we didn't do a unified API for `configure_sets`, meaning there were two different APIs:
+
+```rust
+app.configure_set(Update, A.after(B));
+app.configure_sets(Update, (A.after(B), B.after(C));
+```
+
+In **Bevy 0.12**, we have unified these under a single API to align with the patterns we've used elsewhere and cut down on unnecessary API surface:
+
+```rust
+app.configure_sets(Update, A.after(B));
+app.configure_sets(Update, (A.after(B), B.after(C));
+```
+
+## UI Materials
+
+<div class="release-feature-authors">authors: @MarkusTheOrt</div>
+
+Bevy's material system has been brought to Bevy UI thanks to the new [`UiMaterial`]:
+
+![ui material](ui_material.png)
+
+This "circle" UI Node is drawn with a custom shader:
+
+```rust
+#import bevy_ui::ui_vertex_output::UiVertexOutput
+
+struct CircleMaterial {
+    @location(0) color: vec4<f32>
+}
+
+@group(1) @binding(0)
+var<uniform> input: CircleMaterial;
+
+@fragment
+fn fragment(in: UiVertexOutput) -> @location(0) vec4<f32> {
+    let uv = in.uv * 2.0 - 1.0;
+    let alpha = 1.0 - pow(sqrt(dot(uv, uv)), 100.0);
+    return vec4<f32>(input.color.rgb, alpha);
+}
+```
+
+And just like other Bevy material types, it is simple to set up in code!
+
+```rust
+#[derive(AsBindGroup, Asset, TypePath, Debug, Clone)]
+struct CircleMaterial {
+    #[uniform(0)]
+    color: Vec4,
+}
+
+impl UiMaterial for CircleMaterial {
+    fn fragment_shader() -> ShaderRef {
+        "shaders/circle_shader.wgsl".into()
+    }
+}
+
+// Register the material plugin in your App
+app.add_plugins(UiMaterialPlugin::<CircleMaterial>::default())
+
+// Later in your app, spawn the UI node with your material!
+commands.spawn(MaterialNodeBundle {
+    style: Style {
+        position_type: PositionType::Absolute,
+        width: Val::Px(250.0),
+        height: Val::Px(250.0),
+        ..default()
+    },
+    material: materials.add(CircleMaterial {
+        color: Color::WHITE.into(),
```

Comment:
Not entirely sure if this is a bug or something the shader does, but why is the circle green if the color input is white?

Reply:
it should be white with just this extract. the full example has a system in update that changes the color every frame